### PR TITLE
Refactoring code and adding support for DRMM

### PR DESCRIPTION
**Pull request by pushkalkatara**
Added seperate data preprocessing steps for each model.

Defined each model data preprocessing scripts and params seperately.

### CODE REVIEW

**Comment by pushkalkatara** [comment]
Repetetive file found - https://github.com/arita37/mlmodels/pull/227/files#diff-318b33cb4eec0ee9c658e5ed3e8a2590

**Comment by arita37** [comment]
What about if we try other dataset than wikipedia ???

Can you check :
https://github.com/arita37/mlmodels/blob/dev/mlmodels/model_gluon/gluonts_model.py

**Comment by pushkalkatara** [comment]
In the seperate files, the only change is the data preprocessing scripts and model params. In the case of gluon models the data preprocessing steps and params are not applied as in matchzoo models.

If we can somehow find an alternative for preprocessing the data, we can remove the files for individual model and directly send the data to the model. Also, for different models, the data preprocessing scripts are different.

For other dataset, we can easily specify its name in the JSON. The sciprt would load that dataset if its available in matchzoo and send the raw dataset to individual model data preprocessing scripts for the current implementation.

File: mlmodels/model_tch/matchzoo_bert.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+        no_decay = ['bias', 'LayerNorm.weight']
+        optimizer_grouped_parameters = [
+            {'params': [p for n, p in self.named_parameters() if not any(nd in n for nd in no_decay)], 'weight_decay': 5e-5},
+            {'params': [p for n, p in self.named_parameters() if any(nd in n for nd in no_decay)], 'weight_decay': 0.0}
+        ]
+        return optimizer_grouped_parameters
+
+    def prepare_data(self, data_pars, train_pack_raw, dev_pack_raw, test_pack_raw):
+        preprocessor = self.get_default_preprocessor()
+        train_pack_processed = preprocessor.transform(train_pack_raw)
```
Comment (pushkalkatara) [inline mlmodels/model_tch/matchzoo_bert.py:30]: Here .transform() is used.

File: mlmodels/model_tch/matchzoo_drmm.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        self.params['hist_bin_size'] = 30
+        self.params['mlp_num_layers'] = 1
+        self.params['mlp_num_units'] = 10
+        self.params['mlp_num_fan_out'] = 1
+        self.params['mlp_activation_func'] = 'tanh'
+        self.build()
+        
+    def prepare_data(self, data_pars, train_pack_raw, dev_pack_raw, test_pack_raw):
+        preprocessor = self.get_default_preprocessor()
+        train_pack_processed = preprocessor.fit_transform(train_pack_raw)
```
Comment (pushkalkatara) [inline mlmodels/model_tch/matchzoo_drmm.py:27]: Here .fit_transform() is used.

File: mlmodels/model_tch/matchzoo_drmm.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        dev_pack_processed = preprocessor.transform(dev_pack_raw)
+        test_pack_processed = preprocessor.transform(test_pack_raw)
+
+        glove_embedding = mz.datasets.embeddings.load_glove_embedding(dimension=300)
```
Comment (pushkalkatara) [inline mlmodels/model_tch/matchzoo_drmm.py:31]: Glove embedding is used.

File: mlmodels/model_tch/matchzoo_classification_bert.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    
+    def prepare_data(self, data_pars, train_pack_raw, dev_pack_raw, test_pack_raw):
+        preprocessor = self.get_default_preprocessor()
+        train_pack_processed = preprocessor.transform(train_pack_raw)
+        dev_pack_processed = preprocessor.transform(dev_pack_raw)
+        test_pack_processed = preprocessor.transform(test_pack_raw)
+
+        train_mode = data_pars["mode"] if "mode" in data_pars else None
+        train_sort = data_pars["train_sort"] if "train_sort" in data_pars else False
+
```
Comment (pushkalkatara) [inline mlmodels/model_tch/matchzoo_classification_bert.py:28]: Glove embedding not used

File: mlmodels/model_tch/matchzoo_drmm.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+        term_index = preprocessor.context['vocab_unit'].state['term_index']
+        embedding_matrix = glove_embedding.build_matrix(term_index)
+        l2_norm = np.sqrt((embedding_matrix * embedding_matrix).sum(axis=1))
+        self.embedding_matrix = embedding_matrix / l2_norm[:, np.newaxis]
+
+        histgram_callback = mz.dataloader.callbacks.Histogram(
```
Comment (pushkalkatara) [inline mlmodels/model_tch/matchzoo_drmm.py:37]: New callback, historgram_callback is used, 

This is not used in BERT data preprocessing steps.

File: mlmodels/model_tch/matchzoo_bert.py [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+        train_sort = data_pars["train_sort"] if "train_sort" in data_pars else False
+        trainset = mz.dataloader.Dataset(
+            data_pack=train_pack_processed,
+            mode=train_mode,
+            num_dup=train_num_dup,
+            num_neg=train_num_neg,
+            batch_size=data_pars["train_batch_size"],
+            resample=True,
+            sort=False,
+        )
```
Comment (pushkalkatara) [inline mlmodels/model_tch/matchzoo_bert.py:47]: histogram_callback() not used.

**Pull request closed without merging**